src/senaite/core/adapters/localroles.py

```diff
@@ -0,0 +1,89 @@
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright 2018-2022 by it's authors.
+# Some rights reserved, see README and LICENSE.
+
+import six
+from bika.lims import api
+from bika.lims import logger
+from borg.localrole.default_adapter import DefaultLocalRoleAdapter
+from senaite.core.interfaces import IDynamicLocalRoles
+from zope.component import getAdapters
+
+
+class DynamicLocalRoleAdapter(DefaultLocalRoleAdapter):
+    """Gives additional Member local roles based on current user and context
+    This enables giving additional permissions on items out of the user's
+    current traverse path
+    """
+
+    _roles_in_context = {}
+
+    def getRolesInContext(self, context, principal_id):
+        """Returns the dynamically calculated 'local' roles for the given
+        principal and context
+        @param context: context to calculate roles for the given principal
+        @param principal_id: User login id
+        @return List of dynamically calculated local-roles for user and context
+        """
+        if not api.is_object(context):
+            # We only apply dynamic local roles to valid objects
+            return []
+
+        # This function is called a lot within same request, do some cache
+        context_uid = api.get_uid(context)
+        roles = self._roles_in_context.get(context_uid, {})
+        if principal_id in roles:
+            return roles.get(principal_id)
+
+        # Look for adapters
+        roles = []
+        path = api.get_path(context)
+        adapters = getAdapters((context,), IDynamicLocalRoles)
+        for name, adapter in adapters:
+            local_roles = adapter.getRoles(principal_id)
+            logger.info(u"{}::{}::{}: {}".format(name, path, principal_id,
+                                                 repr(local_roles)))
+            roles.extend(local_roles)
+
+        # Store in cache
+        self._roles_in_context.setdefault(context_uid, {}).update({
+            principal_id: list(set(roles))
+        })
+        return self._roles_in_context[context_uid][principal_id]
+
+    def getRoles(self, principal_id):
+        """Returns both non-local and local roles for the given principal in
+        current context
+        @param principal_id: User login id
+        @return: list of non-local and local roles for the user and context
+        """
+        default_roles = self._rolemap.get(principal_id, [])
+        dynamic_roles = self.getRolesInContext(self.context, principal_id)
+        return list(set(default_roles + dynamic_roles))
+
+    def getAllRoles(self):
+        roles = {}
+        # Iterate through all members to extract their dynamic local role for
+        # current context
+        mtool = api.get_tool("portal_membership")
+        for principal_id in mtool.listMemberIds():
+            user_roles = self.getRoles(principal_id)
+            if user_roles:
+                roles.update({principal_id: user_roles})
+        return six.iteritems(roles)
```

Comment:
The original implementation looks like this:

```python
def getAllRoles(self):
    """Returns all the local roles assigned in this context:
    (principal_id, [role1, role2])"""
    return six.iteritems(self._rolemap)
```
So as far as I understand, this method should operate only on the context and not globally over all members by looking at `__ac_local_roles__` to find local roles stored persistently on an object.

Reply:
Ah ok, I didn't know that. Because the dynamic local roles aren't stored inside the attribute `__ac_local_roles__` we need to go the other way around and lookup the roles from all users, so that our adapter is called again, right?

Reply:
> Because the dynamic local roles aren't stored inside the attribute `__ac_local_roles__` we need to go the other way around and lookup the roles from all users, so that our adapter is called again, right?

Exactly!